File: leet_BTREE_0894_find_all_full_tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func allPossibleFBT(n int) []*TreeNode {
	if n == 1 {
		return []*TreeNode{ &TreeNode {}}
	}
	if n%2 == 0 {
		return []*TreeNode {}
	}

	var res []*TreeNode
	i := 1
	for i < n - 1 {
		L := allPossibleFBT(i)
		R := allPossibleFBT(n - i - 1)
		for _, l := range L {
			for _, r := range R {
				node := & TreeNode{}
				node.Left = l
				node.Right = r
				res = append(res, node)
			}
		}
		i += 2
	}
	return res
}

File: leet_BTREE_0894_find_all_full_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def allPossibleFBT(self, n: int) -> List[Optional[TreeNode]]:
        if n == 1:
            return [ TreeNode() ]
        if n % 2 == 0: # no full tree is possible
            return []

        res = []
        i = 1
        # print('here')
        while i < n - 1:
            # print('here2')
            L = self.allPossibleFBT(i)
            R = self.allPossibleFBT(n - i - 1)
            for l in L:
                # print('here3')
                for r in R:
                    node = TreeNode() # root of a new subtree
                    node.left = l
                    node.right = r
                    res.append(node)
            i += 2
        return res

